### PR TITLE
[FEA] support StringTranslate function

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
We add support for StringTranslate function.

`translate(input, from, to)` - Translates the input string by replacing the characters present in the `from` string with the corresponding characters `in` the to string.

For instance,
```
scala> val translate_df = df.withColumn("c2", translate($"c1", "abc", "123"))
translate_df: org.apache.spark.sql.DataFrame = [c1: string, c2: string]

scala> translate_df.show
+------+------+
|    c1|    c2|
+------+------+
|AaBbCc|A1B2C3|
|  null|  null|
+------+------+
```

### CODE REVIEW

**Review by revans2 — COMMENTED** [review]
Looking good. Need to generate the update docs too.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -819,6 +819,90 @@ case class GpuStringReplace(
...
+      fromExpr: GpuScalar,
+      toExpr: GpuScalar): ColumnVector = {
+    // When from or to string is null, return all nulls like the CPU does.
+    if (!fromExpr.isValid || !toExpr.isValid) {
+      GpuColumnVector.columnVectorFromNull(strExpr.getRowCount.toInt, StringType)
+    } else if (fromExpr.getValue.asInstanceOf[UTF8String].numChars() == 0) {
+      // Return original string if search string is empty
+      strExpr.getBase.asStrings()
+    } else {
+      val fromCharsList = fromExpr.getValue.asInstanceOf[UTF8String].toString.sliding(1).toList
```
[the remark this answers is not in the document]
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:897]: Thank you! I take a look at the Spark code [`srcEval.asInstanceOf[UTF8String].translate(dict)`](https://github.com/apache/spark/blob/fb9d706784557ef0fe9e17d59b7096374658954e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L879). Our source expression is a `GpuColumnVector`. I am not sure if I can use the UTF8String `translate` function as it is in Spark. Please let me know if I understood this incorrectly.

[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:897]: My point was to look at how `dict` was made. 

https://github.com/apache/spark/blob/56043c24a55663e692c3671ba263e642aebc1a50/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L833-L859

java uses UCS2 which is a 16-bit character representation internally. So any unicode character that needs more than 16-bits to represent it has a multi-char/short form in UCS2 (not really sure what the equivalent of multi-byte is when the char is a short).  All of the characters that you tested with fit in a 16-bit representation. The `buildDict` is explicitly setup to deal with these types of characters. It calls `charCount` and `codePointAt` to split the input string apart properly.

[the remark this answers is not in the document]
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala]: Thank you so much! Changes updated.

**Comment by cindyyuanjiang** [comment]
build

File: integration_tests/src/main/python/string_test.py [hunk 2]
```diff
@@ -430,6 +430,19 @@ def test_replace():
                 'REPLACE(a, NULL, "PROD")',
                 'REPLACE(a, "T", "")'))
 
+def test_translate():
+    gen = mk_str_gen('.{0,5}TEST[\ud720 A]{0,5}')
+    assert_gpu_and_cpu_are_equal_collect(
+            lambda spark: unary_op_df(spark, gen).selectExpr(
+                'translate(a, "TEST", "PROD")',
```
[the remark this answers is not in the document]
Comment (revans2) [reply integration_tests/src/main/python/string_test.py:437]: Technically all of these unicode characters still fit in 16-bits. Anything with a 4 hex values will fit in 16 bits.

https://realpython.com/python-encodings-guide/#one-byte-two-bytes-three-bytes-four might help with getting some of these into your python tests.

**Review by revans2 — DISMISSED** [review]
Looks good to me. Just a nit on the testing

**Comment by cindyyuanjiang** [comment]
> Looks good to me. Just a nit on the testing

Thank you! I have some new observation during further testing with large unicode characters (> 16bits). For example,

`a = 'Þp¢{ÍTEST\U0010ffff\U0010ffff A '`
For CPU: `Row(translate(a, T\U0010ffff, PROD)='Þp¢{ÍPESPRORO A '`
For GPU: `Row(translate(a, T\U0010ffff, PROD)='Þp¢{ÍPESPRR A '`

CPU Spark is translating '\U0010ffff' into two characters. While we follow the same mechanism for building the character lists, GPU is translating into 1 character. I am still looking into how to make the results consistent.

**Comment by cindyyuanjiang** [comment]
build

**Comment by sameerz** [comment]
build

**Review by NVnavkumar — COMMENTED** [review]
Not entirely sure, but I wonder if we should be testing and emulating the large codepoint bug in Spark 3.1 versions here. 

Implementation otherwise LGTM aside from minor nits

File: integration_tests/src/main/python/string_test.py [hunk 3]
```diff
@@ -430,6 +430,26 @@ def test_replace():
...
+                'translate(a, "TEST", "P")',
+                'translate(a, "T\ud720", "PROD")',
+                'translate(a, "", "PROD")',
+                'translate(a, NULL, "PROD")',
+                'translate(a, "TEST", NULL)',
+                'translate("AaBbCc", "abc", "123")',
+                'translate("AaBbCc", "abc", "1")'))
+
+@pytest.mark.skipif(is_before_spark_320(), reason="https://issues.apache.org/jira/browse/SPARK-34094")
```
[the remark this answers is not in the document]
Comment (cindyyuanjiang) [reply integration_tests/src/main/python/string_test.py]: Thank you! Changes updated.

**Comment by cindyyuanjiang** [comment]
build

**Review by revans2 — APPROVED** [review]
nit: would be nice to document that we are compatible with 3.2+ Spark for this and because it was a bug fix we have kept it in all versions. But this is super minor.

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by NVnavkumar — APPROVED** [review]

**Comment by cindyyuanjiang** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 4]
```diff
@@ -2956,6 +2956,23 @@ object GpuOverrides extends Logging {
...
+        override def convertToGpu(
+            input: Expression,
+            from: Expression,
+            to: Expression): GpuExpression =
+          // Compatible with Spark 3.2+. Only in Spark 3.2+ does translate() support
+          // unicode characters with code point >= U+10000. Since it was a bug fix,
+          // we have kept it in all versions.
+          // See https://issues.apache.org/jira/browse/SPARK-34094.
+          GpuStringTranslate(input, from, to)
+      }),
```
Comment (NVnavkumar) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2975]: You should switch this to using `incompat` to note the compatibility issue on the generated online documentation. Then regenerate the documentation. Something like the following

```suggestion

          GpuStringTranslate(input, from, to)
      }).incompat("The GPU supports all unicode code points. In Spark versions < 3.2.0 translate() does not " +  "support unicode characters with code point < U+10000 (See SPARK-34094)"),
```

Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala]: Thank you! Changes updated.

**Comment by cindyyuanjiang** [comment]
build

**Comment by cindyyuanjiang** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 5]
```diff
@@ -2956,6 +2956,21 @@ object GpuOverrides extends Logging {
               target: Option[Expression] = None): GpuExpression =
             GpuStringTrimRight(column, target)
         }),
+    expr[StringTranslate](
+      "StringTranslate operator",
```
Comment (NVnavkumar) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2960]: This should just be the definition of the function. You can just use the copy from here: https://spark.apache.org/docs/2.3.0/api/sql/index.html#translate

**Review by NVnavkumar — COMMENTED** [review]
minor documentation nit

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 5 above)
Comment (gerashegalov) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2960]: Looks like we are pretty inconsistent in what we put into `desc`. Not for this PR, but I wonder if at least for *functions* we should just use permalinks to the Spark SQL doc?

**Pull request merged**